b2g/components/RemoteControlService.jsm

```diff
@@ -0,0 +1,395 @@
+
+    Services.obs.removeObserver(this, "xpcom-shutdown");
+
+    if (this._mDNSCancelableHandler) {
+      this._mDNSCancelableHandler.Cancel(Cr.NS_OK);
+      this._mDNSCancelableHandler = null;
+    }
+
+    this._socket.close();
+    this._socket = null;
+    this._doQuit = false;
+    this._serverStatus = SERVER_STATUS.STOPPED;
+
+    return true;
+  },
+
+  // Observers and Listeners
+  // nsIObserver
+  observe: function(subject, topic, data) {
+    switch (topic) {
+      case "xpcom-shutdown": {
+        // Stop service when xpcom-shutdown
+        this.stop();
+        break;
+      }
+    }
+  },
+
+  // nsIServerSocketListener
+  onSocketAccepted: function(socket, trans) {
+    DEBUG && debug("onSocketAccepted(socket=" + socket + ", trans=" + trans + ")");
+    DEBUG && debug("New connection on " + trans.host + ":" + trans.port);
+
+    const SEGMENT_SIZE = 8192;
+    const SEGMENT_COUNT = 1024;
+
+    try {
+      var input = trans.openInputStream(0, SEGMENT_SIZE, SEGMENT_COUNT)
+                       .QueryInterface(Ci.nsIAsyncInputStream);
+      var output = trans.openOutputStream(0, 0, 0);
+    } catch (e) {
+      DEBUG && debug("Error opening transport streams: " + e);
+      trans.close(Cr.NS_BINDING_ABORTED);
+      return;
+    }
+
+    let connectionNumber = ++this._connectionGen;
+
+    try {
+      // Create a connection for each user connection
+      // EventHandler implements nsIInputStreamCallback for incoming message from user
+      var conn = new Connection(input, output, this, socket.port, trans.port, connectionNumber);
+      let handler = new EventHandler(conn);
+
+      input.asyncWait(handler, 0, 0, Services.tm.mainThread);
+    } catch (e) {
+      DEBUG && debug("Error in initial connection: " + e);
+      trans.close(Cr.NS_BINDING_ABORTED);
+      return;
+    }
+
+    this._connections.set(connectionNumber, conn);
+    DEBUG && debug("Start connection " + connectionNumber);
+  },
+
+  // Close all connection when socket closed
+  onStopListening: function(socket, status) {
+    DEbug && debug("Shut down server on port " + socket.port);
+
+    this._connections.forEach(function(value, key){
+      let connection = value;
+      connection.close();
+    });
+
+    this._socketClosed = true;
+  },
+
+  // PRIVATE FUNCTIONS
+  _doStart: function(aResolve, aReject) {
+    DEBUG && debug("doStart");
+
+    if (this._socket) {
+      aReject("SocketAlreadyInit");
+      return;
+    }
+
+    this._doQuit = this._socketClosed = false;
+
+    // Monitor xpcom-shutdown to stop service and clean up
+    Services.obs.addObserver(this, "xpcom-shutdown", false);
+
+    // Start TLSSocketServer with self-signed certification
+    Cc["@mozilla.org/psm;1"].getService(Ci.nsISupports);
+    certService.getOrCreateCert("RemoteControlService", {
+      handleCert: function(cert, result) {
+        if(result) {
+          aReject("getOrCreateCert " + result);
+        } else {
+          let self = RemoteControlService;
```

Comment:
You should move this line out of `getOrCreateCert` block and do it like `let self = this;`